text/0000-pin_and_move.md

```diff
@@ -0,0 +1,381 @@
+```
+
+For types which implement `Move`, `Pin` is essentially the same as an `&mut T`.
+But for types which do not, the conversion between `&mut T` and `Pin` is
+unsafe (however, `Pin` can be easily immutably dereferenced, even for `!Move`
+types).
+
+The contract on the unsafe part of `Pin`s API is that a Pin cannot be
+constructed if the data it references would ever move again, and that it cannot
+be converted into a mutable reference if the data might ever be moved out of
+that reference. In other words, if you have a `Pin` containing data which does
+not implement `Move`, you have a guarantee that that data will never move.
+
+The next two subsections describe safe APIs for constructing a `Pin` of data
+which cannot be moved - one in the heap, and one in the stack.
+
+### Pinning to the heap: The `Anchor` type
+
+The `Anchor` wrapper takes a type that implements `StableDeref` and `Own`, and
+prevents users from moving data out of that unless it implements `Move`. It is
+added to `std::mem` and `core::mem`.
+
+```rust
+struct Anchor<T> {
+     ptr: T,
+}
+
+impl<T: StableDerefMut + Own> Anchor<T> {
+     pub fn new(ptr: T) -> Anchor<T>;
+
+     pub unsafe fn get_mut(this: &mut Anchor<T>) -> &mut T;
+
+     pub unsafe fn into_inner_unchecked(this: Anchor<T>) -> T;
+
+     pub fn pin<'a>(this: &'a mut Anchor<T>) -> Pin<'a, T::Target>;
+}
+
+impl<T: StableDerefMut + Own> Anchor<T> where T::Target: Move {
+     pub fn into_inner(this: Anchor<T>) -> T;
+}
+
+impl<T: StableDerefMut + Own> Deref for Anchor<T> {
+     type Target = T;
+}
+
+impl<T: StableDerefMut + Own> DerefMut for Anchor<T> where T::Target: Move { }
+```
+
+Because `Anchor` implements `StableDeref` and `Own`, and it is not safe to get
+an `&mut T` if the target of `T ` does not implement `Move`, an anchor
+guarantees that the target of `T` will never move again. This satisfies the
+safety constraints of `Pin`, allowing a user to construct a `Pin` from an
+anchored pointer.
+
+Because the data is anchored into the heap, you can move the anchor around
+without moving the data itself. This makes anchor a very flexible way to handle
+immovable data, at the cost of a heap allocation.
+
+An example use:
+
+```rust
+let mut anchor = Anchor::new(Box::new(immovable_data));
+let pin = Anchor::pin(&mut anchor);
+```
+
+### Pinning to the stack: `Pin::stack` and `pinned`
+
+Data can also be pinned to the stack. This avoids the heap allocation, but the
+pin must not outlive the data being pinned, and the API is less convenient.
+
+First, the pinned function, added to `std::mem` and `core::mem`:
+
+```rust
+pub struct StackPinned<'a, T: ?Sized> {
+     _marker: PhantomData<&'a mut &'a ()>,
+     data: T
+}
+
+pub fn pinned<'a, T>(data: T) -> StackPinned<'a, T> {
+     StackPinned {
+          data, _marker: PhantomData,
+     }
+}
+```
+
+Second, this constructor is added to `Pin`:
+
+```rust
+impl<'a, T: ?Sized> Pin<'a, T> {
+    pub fn stack(data: &'a mut StackPinned<'a, T>) -> Pin<'a, T>;
```

Comment:
Similar question to above: what's the rationale for having this be a function on `Pin` rather than a method on `StackPinned`? When [RFC 66](https://github.com/rust-lang/rust/issues/15023) is fully implemented, we could write `mem::pinned(data).stack()`, which looks nicer to me than the two-liner below.

Reply:
Nit: the names of the functions are a bit misleaing here, too. `pinned` gives you a `StackPinned`, which isn't actually pinned to the stack-- you're free to write `Box::new(mem::pinned(data))`. What it does do is move your `T` and tie it to an invariant lifetime `'a`. The `stack` function is what actually prevents the `StackPinned` from ever being moved out of its current location-- it's the thing that actually does the pinning.

I don't have any great suggestions here. Maybe `LifetimePinned` instead of `StackPinned`?

It's unfortunate how much of the complexity in these APIs winds up being passed on to the end user, especially including all of the lifetime annotations. It might be that an inbuilt `&pin` reference (which binds to the stack for `!Move` types, and is equal to `&mut` for `Move` types) would be less scary.

Reply:
> It's unfortunate how much of the complexity in these APIs winds up being passed on to the end user, especially including all of the lifetime annotations. It might be that an inbuilt &pin reference (which binds to the stack for !Move types, and is equal to &mut for Move types) would be less scary.

Importantly, the library version should be forward compatible with the language version.

(I also think stack pinning is a niche use case compared to heap pinning though).

Reply:
> (I also think stack pinning is a niche use case compared to heap pinning though).

I agree-- I think that `&pin` could allow the heap-pinning case to be more ergonomic: rather than `my_anchored_boxed_future.as_pin().poll(cx)`, `&pin` could allow writing `my_anchored_boxed_future.poll(cx)`, as `Future::poll` would take `&pin self`, rather than `self: Pin<'a, Self>`.

Reply:
That's  true. This RFC should be forward compatible with doing that someday by just changing Pin to this:

```rust
type Pin<'a, T> = &'a pin T;
```

Reply:
Rather than just renaming `StackPinned` to `LifetimePinned`, I think the `*Pinned` type is the one that should be named `Pin`, and the `Pin` one should be named `*Pinned`.

You put something in a `Pin` which you can then move, and `Pin::pin` then return a `Pinned` pointing to a value that can't be moved anymore.

I hope that makes sense.